### ISSUE TITLE
Significant performance drop when training from scratch on a single GPU (RTX 4090)

### ISSUE THREAD

**Issue by RyanPham19092002** [Tsinghua-MARS-Lab/TrackOcc#7]
Hi @zgchen33 ,

First of all, thank you so much for sharing your code! This is truly a great piece of work and a very valuable contribution to the community.

I am currently trying to re-implement and train the model from scratch using a single RTX 4090 GPU. To adapt to my hardware, I modified the configuration by changing `num_gpus = 8` down to `1`. The rest of the codebase and configurations remain entirely unchanged.

I started the training process using the following command:
`./dist_train.sh configs/TrackOcc/trackocc_r50_704x256_3inst_3f_8gpu.py 1`

After training for more than 10 days, I evaluated the saved EMA checkpoint (`work_dirs/trackocc_r50_704x256_3inst_3f_8gpu test 2/iter_758807_ema.pth`). However, the evaluation metrics are significantly lower compared to the pretrained weights provided on Hugging Face. 

Specifically:
- **Evaluation from your provided Pretrained weights:** `{'panoptic4d': (0.2009190578613111, 0.13714387456068158, 0.2943512274331667)}`

<img width="1080" height="147" alt="Image" src="https://github.com/user-attachments/assets/a72da6e9-4f19-4f08-bb5b-5a1b38b85fe3" />

- **Evaluation from my Re-implementation (1 GPU):** `{'panoptic4d': (0.15886513658199553, 0.09196984863850512, 0.27441745305482235)}`

<img width="1077" height="148" alt="Image" src="https://github.com/user-attachments/assets/c543f979-ee11-4072-bc06-24d5e9d9d2ff" />

Since the effective batch size was reduced by 8 times (from 8 GPUs to 1 GPU), I suspect this might be the main cause of the performance drop. 

Could you kindly provide some guidance on how I should adjust the configuration (e.g., modifying the learning rate, using gradient accumulation, etc.) to achieve comparable results when training on a single GPU? 

Thank you again for your time and your help!

**Comment by zgchen33** [comment]
Yes, the reduction in effective batch size (from 8 GPUs to a single GPU) could very likely be a major factor contributing to the performance drop.

To mitigate this, you may consider the following adjustments:

- Gradient accumulation: This is a common approach to simulate a larger effective batch size when limited by GPU resources. However, since this codebase is relatively old, it may not have built-in support for gradient accumulation, so you would likely need to implement it manually.

- Learning rate adjustment: You could also try tuning the learning rate accordingly. In practice, I usually scale the learning rate proportional to the square root of the effective batch size. Since your batch size is reduced by a factor of 8, you might try reducing the learning rate by a factor of √8 and see if that helps stabilize performance.

These adjustments may help you achieve results closer to the multi-GPU setting.

**Comment by RyanPham19092002** [comment]
Hi @zgchen33 ,

Thank you so much for the helpful suggestions regarding the effective batch size!

Following your advice, I have tested both approaches to mitigate the gap between the single-GPU and 8-GPU setups. Unfortunately, the performance is still significantly lower than expected in both cases.

Here is the breakdown of the experiments I ran:

**Experiment 1: Gradient Accumulation**
I manually implemented gradient accumulation to simulate the 8-GPU effective batch size, keeping the original learning rate at 2e-4.

**Config I modified:
optimizer_config = dict(
    cumulative_iters=8, grad_clip=dict(max_norm=35, norm_type=2))**

- Result summary: PQ4D = 0.169, AQ_overall = 0.109, iou_mean = 0.262
<img width="1180" height="159" alt="Image" src="https://github.com/user-attachments/assets/e60c195e-edbd-4024-bc14-eefabef6b2af" />

**Experiment 2: Learning Rate Adjustment**
I ran the model on a single GPU (without gradient accumulation) and scaled down the learning rate by a factor of sqrt(8), changing it from 2e-4 to 5e-5.

- Result summary: PQ4D = 0.161, AQ_overall = 0.106, iou_mean = 0.243
<img width="1189" height="161" alt="Image" src="https://github.com/user-attachments/assets/4f467d79-6c03-4697-9a24-064c1b93084f" />

Given that these two adjustments didn't fully bridge the gap, I'm wondering if there are other aspects of the training process I need to adapt for a single-GPU setup. Could you suggest any other methods, hyperparameter changes, or specific tricks that might help me achieve results closer to the original multi-GPU setting? Any further guidance would be greatly appreciated!

Thanks again for your time!
